### PR TITLE
Fix: Batch posting delayed because of the frequency of messages

### PR DESCRIPTION
**Pull request by Sneh1999**
Closes https://github.com/EspressoSystems/nitro-espresso-integration/issues/439

### CODE REVIEW

File: arbnode/batch_poster.go [hunk 1]
```diff
@@ -1322,11 +1322,30 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
...
+  if !b.building.haveUsefulMessage{ // If we don't have a useful message we need to check if we are at the batch limit for this iteration and delete the upper bound
+                                    // and allow at least 1 useful message to flow through.  
+      if atEspressoBatchLimit{
+        // if we are at our msg limit for the batch and we don't have a useful message, we have a 1 message batch with just the batch posting report
+        // if this is the case, we should delete the upper bound and process some more messages, otherwise we don't have a useful batch and forcePostBatch = true is useless
+        err = b.deleteLastPotentialMsgInBatch()
+        if err != nil{
+          return false, err
+        }
+        log.Info("This current batch was full, but did not have a useful message. We should delete the espresso upper bound on the batch")
```
Comment (jbearer) [inline arbnode/batch_poster.go:1337]: Is this possible? The batch can be full but not have a useful message? What makes a message "useful"?

File: arbnode/batch_poster.go [hunk 2]
```diff
@@ -1322,11 +1322,30 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
 		return false, nil
 	}
 
-	lastPotentialMsg, err := b.streamer.GetMessage(msgCount - 1)
+
+	lastPotentialMsg, lastPotentialMsgInBatchPos, err := b.getLastPotentialMsgInBatch(msgCount)
+  atEspressoBatchLimit := b.building.msgCount >= *lastPotentialMsgInBatchPos
```
Comment (jbearer) [inline arbnode/batch_poster.go:1327]: Potential `nil` dereference?

File: arbnode/transaction_streamer.go [hunk 3]
```diff
@@ -1388,6 +1388,42 @@ func (s *TransactionStreamer) getEspressoSubmittedPos() ([]arbutil.MessageIndex,
...
+	var msgWithMetadata arbostypes.MessageWithMetadata
+	err = rlp.DecodeBytes(lastPotentialMsgBytes, &msgWithMetadata)
+
+	if err != nil {
+		return nil, err
+	}
+	return &msgWithMetadata, nil
+}
+
+func (s *TransactionStreamer) getLastPotentialMsgPos() (*arbutil.MessageIndex, error) {
```
Comment (jbearer) [inline arbnode/transaction_streamer.go:1409]: Is this necessary? We don't have the message pos in the metadata of `MessageWithMetadata` returned from `getLastPotentialMsg`? If so then fine, it's just a little sad

File: arbnode/batch_poster.go (hunk 2 above)
Comment (jbearer) [inline arbnode/batch_poster.go:1327]: I find this confusing. `getLastPotentialMsgInBatch` sets the last potential message if it is not set. So, after the first time this is called, `lastPotentialMsgInBatchPos` will always be set. Then, we cut off building of this batch at that `lastPotentialMsgInBatchPos`. This means, I think, that the batch will be cut off at whatever messages we have the first time this function is called, even if it's not otherwise time to close the batch.

I thought what we wanted to do is add messages to the batch unrestricted as we normally do, and then *only when it is time to close the batch* (ie it's full or the max delay has expired, the conditions that would usually trigger us to post the batch) prohibit the addition of future messages until the current batch is successfully posted.

I feel like I'm misunderstanding so could you explain what this code is trying to do and why it's different (or how it's the same) as what I just described?

File: arbnode/batch_poster.go (hunk 1 above)
Comment (zacshowa) [reply arbnode/batch_poster.go:1337]: Full in this context is referring to the upper bound we place on the batch the first iteration of `maybePostSequencerBatch()` after we have just posted a batch. I believe that the next message after the batch is posted is always a l1 batch posting report which seems to be the only message type that the poster filters out as a not "useful" message [here](https://github.com/EspressoSystems/nitro-espresso-integration/blob/17734f4cfdb58c523aa26870ddf9a57c2f328857/arbnode/batch_poster.go#L1476C3-L1481C4)

In the event that we only have this batch posting report, the poster would enter a state where it puts a 1 message upper bound on the batch, and then includes the batch posting report, but won't have a useful message, and therefore won't post the batch ever as we don't re-enter the for loop that adds messages to the batch. 

Given our current strategy of determining batch size based on the number of messages we have received since we posted the last batch, the poster will rather aggressively limit the number of messages in a batch if it is posting fast which leads to small batches that may just have these l1 batch posting report messages

File: arbnode/batch_poster.go (hunk 2 above)
Comment (zacshowa) [reply arbnode/batch_poster.go:1327]: This code is different than what you describe. This code will cut off the batch at the amount of messages we have the first time we call this function. I can adapt it to only perform this cutoff only after it's time to close the batch.

This will likely lead to the batch being posted some time after we set the message cutoff as we will likely need to wait some time for the validation to happen. I think this wait time will be limited to one posting window from the prover to the light client though. Is that acceptable behavior?

File: arbnode/transaction_streamer.go (hunk 3 above)
Comment (zacshowa) [reply arbnode/transaction_streamer.go:1409]: Unfortunately we [don't](https://github.com/EspressoSystems/nitro-espresso-integration/blob/17734f4cfdb58c523aa26870ddf9a57c2f328857/arbos/arbostypes/messagewithmeta.go#L17C1-L20C2)

File: arbnode/batch_poster.go (hunk 2 above)
Comment (jbearer) [reply arbnode/batch_poster.go:1327]: I think that's better. We want to change the existing behavior of the batcher as little as possible, including only posting batches when they are full (or the max delay has passed). There are important reasons for this behavior such as optimizing the compression rate of the data we post.

The delay between closing the batch and posting it I think is not such a big deal. As you mentioned it should be constant and not accumulate over time, so that's fine. We can always set the max delay to 20 minutes so that we are guaranteed to post within 30.

Comment (zacshowa) [reply arbnode/batch_poster.go]: You're right I should have fixed it in [the commit](https://github.com/EspressoSystems/nitro-espresso-integration/pull/440/commits/4308c308422efff749652195dfd92ea6a53b71a4) to address your other comments.

File: arbnode/batch_poster.go [hunk 4]
```diff
@@ -1409,13 +1409,29 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
...
-
 	for b.building.msgCount < msgCount {
 		msg, err := b.streamer.GetMessage(b.building.msgCount)
 		if err != nil {
 			log.Error("error getting message from streamer", "error", err)
 			return false, fmt.Errorf("error getting message from streamer: %w", err)
 		}
+		// #nosec G115
+		timeSinceMsg := time.Since(time.Unix(int64(msg.Message.Header.Timestamp), 0))
+		if timeSinceMsg >= config.MaxDelay {
```
Comment (jbearer) [inline arbnode/batch_poster.go:1420]: We're looking at time since the latest message. So we good get to say 29 minutes and then look at a message that we just received and decide to include it. Then wait 10 minutes for Espresso finality and by that time we've received another message just now. Even though it's been 40 minutes since we started building the block, this message is new, so we'll include it. Wait 10 more minutes. Etc.

It seems like we should be checking if time since we began building the block >= MaxDelay, not time since the last message.

More generally I feel this would be simpler and easier to understand if we set a cap on the number of messages to be included right before we post. So the logic would be like:

```
if !batchClosed
    accumulate messages
if time to post (block is full or maxdelay has passed)
    batchClosed = true
    if !checkEspressoFinality():
      return and try again
    post batch
```

Something like that. Let me know if it makes sense or you disagree

File: arbnode/batch_poster.go [hunk 5]
```diff
@@ -563,29 +563,29 @@ var EspressoFetchTransactionErr = errors.New("failed to fetch the espresso trans
...
 
 	// This message has passed the espresso verification
-	if lastConfirmed != nil && b.building.msgCount <= *lastConfirmed {
-		return nil
-	}
 
-	if b.streamer.EscapeHatchEnabled {
-		log.Warn("skipped espresso verification due to hotshot failure", "pos", b.building.msgCount)
-		return nil
+	if lastConfirmed != nil && b.building.msgCount-1 <= *lastConfirmed {
```
Comment (zacshowa) [inline arbnode/batch_poster.go:584]: I am a bit worried about this change. With this implementation the last confirmed was always ending one behind `b.building.msgCount` I suspect that this is just an off by one error because the 3.3.2 code also doesn't handle including the message at msgCount in the batch. It also makes me wonder why this check worked in the first place.

I do think we should add some tests around different transaction loads with different config settings for the batch poster so we can at least sniff out some of these weird edge cases with how we build the batch.

File: arbnode/batch_poster.go [hunk 6]
```diff
@@ -1516,7 +1524,11 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
 		// don't post anything for now
 		return false, nil
 	}
-
+	// If we are checking the validation, set isWaitingForEspressoValidation in the batch segments and re-poll the function until we are ready to post.
```
Comment (zacshowa) [inline arbnode/batch_poster.go:1527]: I'm also a bit worried about this with respect to the escape hatch's correctness. The current implementation of the escape hatch decides if hotshot is live based on its most recent header according to the light client.
I think its possible that we could submit some transactions to espresso, then, if liveness has changed, skip verifying them by the time we decide to it's time to post the batch and then check the validation. I'm not sure how we would easily handle this though.

File: arbnode/batch_poster.go (hunk 5 above)
Comment (Sneh1999) [reply arbnode/batch_poster.go:584]: It worked because `checkEspressoValidation` was running in a loop where `b.building.msgCount < msgCount` was holding true with every iteration and `		b.building.msgCount++` was executing at the end of each iteration. So we were essentially checking that with each iteration because the b.building.msgCount is increasing `b.building.msgCount-1 <= *lastConfirmed` last confirmed is always greater or equal. Let me know if that makes sense!

File: arbnode/batch_poster.go [hunk 7]
```diff
@@ -1017,6 +1018,10 @@ func (s *batchSegments) AddMessage(msg *arbostypes.MessageWithMetadata) (bool, e
 	if s.isDone {
 		return false, errBatchAlreadyClosed
 	}
+	if s.isWaitingForEspressoValidation {
```
Comment (Sneh1999) [inline arbnode/batch_poster.go:1021]: can we add a log here? this would help with debugging on devnet

**Comment by zacshowa** [comment]
The only remaining concern I have with this is that after merging with PR's merged to `celestia-integration` today, the E2E test runs for ~120 seconds longer than it did prior to the merge

File: arbnode/batch_poster.go (hunk 6 above)
Comment (Sneh1999) [reply arbnode/batch_poster.go:1504]: I think that is expected behaviour though? if hotshot is not live, we should not wait for espresso validation. CC @jbearer  correct me if my understanding is incorrect here.

File: arbnode/batch_poster.go [hunk 8]
```diff
@@ -1516,7 +1524,11 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
 		// don't post anything for now
 		return false, nil
 	}
-
+	// If we are checking the validation, set isWaitingForEspressoValidation in the batch segments and re-poll the function until we are ready to post.
+  hasBatchBeenValidated:= b.checkEspressoValidation()
+	if !hasBatchBeenValidated {
```
Comment (Sneh1999) [inline arbnode/batch_poster.go:1506]: can we add a log here for debugging purposes of devnet and testnet? we can do log cleanup later but it will help until we have a working implementation

**Comment by Sneh1999** [comment]
> The only remaining concern I have with this is that after merging with PR's merged to `celestia-integration` today, the E2E test runs for ~120 seconds longer than it did prior to the merge

 I added a new test thats why the time might have increased

File: arbnode/batch_poster.go (hunk 5 above)
Comment (jbearer) [reply arbnode/batch_poster.go:589]: So it's just because we switched the order between checking espresso validation and incrementing the counter, and so the old check is equivalent to the new one? That makes sense

Comment (jbearer) [reply arbnode/batch_poster.go:589]: In any case the new one is much more intuitive I think: `b.building.msgCount-1` should be the *position* of the last message in the batch and we are simply checking that this message position is confirmed

File: arbnode/batch_poster.go [hunk 9]
```diff
@@ -1308,7 +1320,8 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
 		// There's nothing after the newest batch, therefore batch posting was not required
 		return false, nil
 	}
-
+	// This will be overwritten when we establish an upper bound on the messages if we take longer than the max delay to post a batch
```
Comment (Sneh1999) [inline arbnode/batch_poster.go:1323]: is this comment relevant now?

File: arbnode/batch_poster.go (hunk 6 above)
Comment (jbearer) [reply arbnode/batch_poster.go:1503]: I think that's not a problem. If the escape hatch has triggered than all bets are off anyways when it comes to Espresso confirmations. In other words the guarantees around Espresso confirmations are always "assuming Espresso remains secure and live". The escape hatch is just there so that the rollup as a whole will continue running if Espresso goes down, but we don't have to worry about the integrity of espresso confirmations in this case

Comment (jbearer) [reply arbnode/batch_poster.go:1503]: Which is exactly why we prefer for rollups not to enable the escape hatch if possible

**Review by jbearer — APPROVED** [review]
